src/core/useAnimations.tsx

```diff
@@ -2,36 +2,48 @@ import * as React from 'react'
 import { Object3D, AnimationClip, AnimationAction, AnimationMixer } from 'three'
 import { useFrame } from '@react-three/fiber'
 
-type Api = {
+type Api<T extends string, U = { [K in T]?: AnimationAction }> = {
   ref: React.MutableRefObject<Object3D | undefined | null>
   clips: AnimationClip[]
   mixer: AnimationMixer
-  names: string[]
-  actions: {
-    [key: string]: AnimationAction
-  }
+  names: T[]
+  actions: U
 }
 
-export function useAnimations(clips: AnimationClip[], root?: React.MutableRefObject<Object3D | undefined | null>) {
+type ExtendedAnimationClip<T extends String> = AnimationClip & { name?: T }
+
+export function useAnimations<T extends string>(
+  clips: ExtendedAnimationClip<T>[],
+  root?: React.MutableRefObject<Object3D | undefined | null>
+): Api<T> {
   const ref = React.useRef<Object3D>()
   const actualRef = root ? root : ref
   const [mixer] = React.useState(() => new AnimationMixer((undefined as unknown) as Object3D))
-  const api: Api = React.useMemo(
-    () => ({
-      ref: actualRef,
-      clips,
-      actions: clips.reduce((prev, clip) => ({ ...prev, [clip.name]: null }), {}),
-      names: clips.map((clip) => clip.name),
-      mixer,
-    }),
-    [clips, mixer, actualRef]
-  )
+  const [api, setApi] = React.useState<Api<T>>({
+    ref: actualRef,
+    clips,
+    actions: clips.map((clip) => clip.name).reduce((prev, name: T) => ({ ...prev, [name]: null }), {}),
+    names: clips.map((clip) => clip.name),
+    mixer,
+  })
   useFrame((state, delta) => mixer.update(delta))
   React.useEffect(() => {
     const currentRoot = actualRef.current
     if (currentRoot) {
-      clips.forEach((clip) => (api.actions[clip.name] = mixer.clipAction(clip, currentRoot)))
+      setApi((oldApi) => ({
```

Comment:
How come you've changed this? I'm a bit out of the loop.

Reply:
Well, I don't really know where the discussion should happen. In the initial issue or in the pull request? But to sum up the discussion over there:
 
 - the API is not returned in the render loop with the actions set (aka every actions are null)
 - I thought that wrapping it with curly brackets fixed the issue but in fact it was just the hot-reload which was fixing the issue
 - However, within the effect loop (aka `useEffect` or `useLayoutEffect`), because we mutate actions directly within `useAnimations`, it's filled and we can use it

So, because it got me confused I suggested here to change the approach for a state one, where the actions could be seen from the component. And I pushed a fixup showing what it could look like for an example.

From my point of view, it makes more sense. But I can see problems with this approach:
 - We add another rendering (the one giving us the actions set)
 - Every bit of code using this hook could now run with the actions not set at first

This means that when we write something like (took from the story book): 
```javascript
  React.useEffect(() => {
    actions[selectedAction].reset().fadeIn(blendDuration).play()
    return () => void actions[selectedAction].fadeOut(blendDuration)
  }, [actions, selectedAction, blendDuration])
```
It would fail because the initial actions is not updated anymore. 
We could also mix both of these strategies (mutate the old one and update the state) so we don't have any breaking change.